cocos2d/core/renderer/DirtyRegion.js

```diff
@@ -0,0 +1,337 @@
+//Region is used to label a rect which world axis aligned.
+var Region = function () {
+    this._minX = 0;
+    this._minY = 0;
+    this._maxX = 0;
+    this._maxY = 0;
+
+    this._width = 0;
+    this._height = 0;
+    this._area = 0;
+    //this.moved = false;
+};
+
+var regionProto = Region.prototype;
+
+var regionPool = [];
+
+function regionCreate() {
+    var region = regionPool.pop();
+    if (!region) {
+        region = new Region();
+    }
+    return region;
+}
+
+function regionRelease(region) {
+    regionPool.push(region);
+}
+
+regionProto.setTo = function (minX, minY, maxX, maxY) {
+    this._minX = minX;
+    this._minY = minY;
+    this._maxX = maxX;
+    this._maxY = maxY;
+    this.updateArea();
+    return this;
+};
+
+//convert region to int values which is fast for clipping
+regionProto.intValues = function () {
+    this._minX = Math.floor(this._minX);
+    this._minY = Math.floor(this._minY);
+    this._maxX = Math.ceil(this._maxX);
+    this._maxY = Math.ceil(this._maxY);
+    this.updateArea();
+};
+
+//update the area of region
+regionProto.updateArea = function () {
+    this._width = this._maxX - this._minX;
+    this._height = this._maxY - this._minY;
+    this._area = this._width * this._height;
+};
+
+//merge two region into one
+regionProto.union = function (target) {
+    if(this.isEmpty()) {
+        this.setTo(target._minX, target._minY, target._maxX, target._maxY);
+        return;
+    }
+    if (this._minX > target._minX) {
+        this._minX = target._minX;
+    }
+    if (this._minY > target._minY) {
+        this._minY = target._minY;
+    }
+    if (this._maxX < target._maxX) {
+        this._maxX = target._maxX;
+    }
+    if (this._maxY < target._maxY) {
+        this._maxY = target._maxY;
+    }
+    this.updateArea();
+};
+
+//regionProto.intersect = function (target) {
+//    if (this._minX < target._minX) {
+//        this._minX = target._minX;
+//    }
+//    if (this._maxX > target._maxX) {
+//        this._maxX = target._maxX;
+//    }
+//    if (this._minX >= this._maxX) {
+//        this.setEmpty();
+//        return;
+//    }
+//    if (this._minY < target._minY) {
+//        this._minY = target._minY;
+//    }
+//
+//    if (this._maxY > target._maxY) {
+//        this._maxY = target._maxY;
+//    }
+//    if (this._minY >= this._maxY) {
+//        this.setEmpty();
+//        return;
+//    }
+//    this.updateArea();
+//};
+
+//set region to empty
+regionProto.setEmpty = function () {
+    this._minX = 0;
+    this._minY = 0;
+    this._maxX = 0;
+    this._maxY = 0;
+    this._width = 0;
+    this._height = 0;
+    this._area = 0;
+};
+
+regionProto.isEmpty = function () {
+    return this._width <= 0 || this._height <= 0;
+};
+
+//check whether two region is intersects or not
+regionProto.intersects = function (target) {
+    if (this.isEmpty() || target.isEmpty()) {
```

Comment:
suggest to avoid using function call, even if the code is longer and less readable, we can save some execution time, I believe intersects will be called many times during each frame.